File: algorithms-and-data-structures/handy-algorithms/split-string.cpp
#include <iostream>
#include <cstring>
#include <vector>
using namespace std;

vector<string> splitAll(const string & s, const string & delim) {
  vector<string> tokens;
  size_t last = 0;
  for(size_t next = 0; (next = s.find(delim, last)) != s.npos; ) {
    if (last < next)
      tokens.push_back(s.substr(last, next - last));
    last = next + delim.size();
  }
  if (last < s.size()) tokens.push_back(s.substr(last));
  return tokens;
}

int main() {
  string s = "  ab cd efg gh   ";
  vector<string> tokens = splitAll(s, " ");
  for(string token :tokens)
  	cout << "'" << token << "'" << '\n';
  return 0;
}

File: algorithms-and-data-structures/handy-algorithms/split-string.java
import java.util.ArrayList;
import java.util.List;

public class Main {

    private static List<String> split(String s, String delim) {
        List<String> tokens = new ArrayList<>();
        int last = 0;
        for(int next; (next = s.indexOf(delim, last)) != -1;) {
            if (last < next) {
                tokens.add(s.substring(last, next));
            }
            last = next + delim.length();
        }
        if (last < s.length()) {
            tokens.add(s.substring(last));
        }
        return tokens;
    }

    public static void main(String[] args) {
    }
}